### ISSUE TITLE
'UseWeatherFile' is not yet a supported option for YearDescription (Bugzilla #979)

### ISSUE THREAD

**Issue by axelstudios** [NREL/OpenStudio#227]
On 2012-11-01 17:07:02, @macumber wrote:

> openstudio::Date YearDescription_Impl::makeDate
> 
> See if we can remove this error when a weather file is set

**Comment by DavidGoldwasser** [comment]
@asparke2 and @mdahlhausen have you noticed this a lot on create typical. This used to just be in console out but now that is written into the out.osw. I'm getting it 5500 times in create_typical measure log vs. only twice in ChangeBuildingLocation.

@tijcolem despite could be good to have someone look at this. Maybe there is change we can make to ruby code and not OpenStudio. I thought we cleaned up some bugs related to logging but not this one.

The full messsage is `[openstudio.model.YearDescription] <1> 'UseWeatherFile' is not yet a supported option for YearDescription`

**Comment by asparke2** [comment]
@DavidGoldwasser Yeah it's gotten out of hand, but haven't looked into where it's coming from yet.  Getting rid of this would be excellent!

**Comment by DavidGoldwasser** [comment]
Sorry @kbenne just adding you for a few that were still assigned to Dan. We can change to someone else I just didn't want it un-assigned.

**Comment by eringold** [comment]
@DavidGoldwasser 
I think it's occurring when the standards gem makes schedules, like: [here](https://github.com/NREL/openstudio-standards/blob/c7e5dc40fa307fc26dbe694ebd4587212faa47ce/lib/openstudio-standards/standards/Standards.ThermalZone.rb#L317) or [here](https://github.com/NREL/openstudio-standards/blob/c04b997f8a8fa847dab7c51172d4f88ac02adb6b/lib/openstudio-standards/standards/Standards.ScheduleRuleset.rb#L262)

My workaround is just to add this right after creating a new model:
`model.getYearDescription.setDayofWeekforStartDay("Sunday")`

**Comment by asparke2** [comment]
Here's my analysis of why this is happening:

1. You have a model where "YearDescription -> Day of Week for Start Day" is blank.  The OS Application sets this field when you assign a .epw file, but when you programmatically build a model you have to do this yourself.

2. You call [`YearDescription::makeDate(some_integer)`](https://github.com/NREL/OpenStudio/blob/develop/src/model/YearDescription.cpp#L253-L275), which generates the message `'UseWeatherFile' is not yet a supported option for YearDescription`

openstudio-standards calls `makeDate()` [a lot here](https://github.com/NREL/openstudio-standards/blob/master/lib/openstudio-standards/standards/Standards.ThermalZone.rb#L295-L296) when used to programatically generate operation schedules for HVAC systems

I think the solution could be to use the [cached version of `Model::assumedYear()`](https://github.com/NREL/OpenStudio/blob/develop/src/model/Model.cpp#L467-L474) inside `YearDescription::makeDate()`.  The message should also be changed to something meaningful like "YearDescription has not been assigned a specific starting day of the week, making dates assuming calendar year 2009" (or whatever the default year is).

**Comment by DavidGoldwasser** [comment]
Thanks @eringold and @asparke2. Quick fix is to use the set_year argument in change building location, which I have been leaving at 0 which doesn't alter it. changing that argument to 2019 got rid of the message :) I still agree there is a bug to fix in API, but nice to know how to get around it.
https://github.com/NREL/openstudio-common-measures-gem/blob/develop/lib/measures/ChangeBuildingLocation/measure.rb#L75-L79

**Comment by joseph-robertson** [comment]
What is the proposed bugfix in the API?

**Comment by jmarrec** [comment]
If you use the integer overload: https://github.com/NREL/OpenStudio/blob/7cec475621233fcd772de2c6715ef9252bcbb7a6/src/model/YearDescription.cpp#L279-L286

It's actually `assumedYear` that issues the warning https://github.com/NREL/OpenStudio/blob/7cec475621233fcd772de2c6715ef9252bcbb7a6/src/model/YearDescription.cpp#L219-L240

Anyways, I think the API change should be to actually correctly implement the UseWeatherFile option.

* EpwFile during parse, will set that Start Day of Week based on the DATA PERIODS line: `EpwFile::startDayOfWeek`
```
DATA PERIODS,1,1,Data,Friday, 1/ 1,12/31
```
* EpwFile during parse, if it's found to be an AMY and not a TMY, will also get the actual year: `EpwFile::startDateActualYear`

I propose to extend the OS:WeatherFile to cache these values, so we don't repeteadly have to actually repase the EpwFile to get this info. So add two fields, and populate them during https://github.com/NREL/OpenStudio/blob/7cec475621233fcd772de2c6715ef9252bcbb7a6/src/model/WeatherFile.cpp#L397

Then in YearDescription::assumedYear, 

```python
if "UseWeatherFile":
  if model has a weather file:
      if weatherfile has an actual start date: 
          use that
      else:
          assumed dow = weatherFile.startDayOfWeek
  else:
    LOG Error: 'UseWeatherFile' but no weather file has been set
else
  // business as usual, assume dow as stored on the YearDescription already
```

Make sure all makeData routine then also call this assumedBaseYear instead of doing their duplicated logic.

**Comment by macumber** [comment]
Closing with an actual fix, hurrah!

:partying_face: :partying_face: :partying_face:

**Comment by asparke2** [comment]
Hooray!

**Comment by jmarrec** [comment]
So, is the action item here to delete the Info messages? Or move them down to a Debug or Trace level?

(I am voluntarily excluding adding state to the YearDescription_Impl class by adding a boolean to record if the message has been issued once since the last time any of the meaningful parameters have been changed. Too complex for no reason)

**Comment by eringold** [comment]
@jmarrec Thanks for the insights about suppressing multiple instances of the message. I think we could handle this by setting the CalendarYear, or setting the channel regex.
